File: 0069. Sqrt(x).py
#Time Complexity: O(sqrt(n))
#Space Complexity: O(1)
#Approach: Iteration
class Solution:
    def mySqrt(self, x: int) -> int:
        i = 1
        while i * i <= x:
            i += 1
        return i - 1


#A faster approach would be with bit manipulation

File: 0069. Sqrt(x).cpp
//Time Complexity: O(logn)
//Space Complexity: O(1)
//Approach: Binary Search, Math
class Solution {
public:
    int mySqrt(int x) {
        int H = 46340; // sqrt(2147483647)
        int L = 0;
        int M = (H + L) / 2;
        while(L <= H){
            if(M * M > x){
                H = M - 1;
            }
            else{
                L = M + 1;
            }
            M = (H + L) / 2;
        }
        return M;
    }
};
